internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,325 @@
+		clusterID := kafkaStreamingUnitCountPerCluster.ClusterId
+		existing, ok := alreadyProcessedClusters[clusterID]
+
+		if !ok {
+			alreadyProcessedClusters[clusterID] = processed{
+				// consider non ready cluster as already processed for scale down evaluation and thus they'll be skipped from scale down consideration
+				processed:                              kafkaStreamingUnitCountPerCluster.Status != api.ClusterReady.String(),
+				indexesOfStreamingUnitForSameClusterID: []int{i},
+			}
+		} else {
+			alreadyProcessedClusters[clusterID] = processed{
+				indexesOfStreamingUnitForSameClusterID: append(existing.indexesOfStreamingUnitForSameClusterID, i),
+				processed:                              existing.processed,
+			}
+		}
+	}
+
+	for _, suCount := range kafkaStreamingUnitCountPerClusterList {
+		clusterID := suCount.ClusterId
+		existing, alreadyProcessed := alreadyProcessedClusters[clusterID]
+		if alreadyProcessed && existing.processed { // skip already processed
+			glog.V(10).Infof("cluster with cluster_id %s is already processed", suCount.ClusterId)
+			continue
+		}
+
+		alreadyProcessedClusters[clusterID] = processed{
+			indexesOfStreamingUnitForSameClusterID: existing.indexesOfStreamingUnitForSameClusterID,
+			processed:                              true,
+		}
+
+		var regionsSupportedInstanceType config.InstanceTypeMap
+		provider, ok := m.clusterProvidersConfig.ProvidersConfig.SupportedProviders.GetByName(suCount.CloudProvider)
+		if ok {
+			region, regionFound := provider.Regions.GetByName(suCount.Region)
+			if regionFound {
+				regionsSupportedInstanceType = region.SupportedInstanceTypes
+			}
+		}
+
+		var dynamicScaleDownProcessor dynamicScaleDownProcessor = &standardDynamicScaleDownProcessor{
+			kafkaStreamingUnitCountPerClusterList:  kafkaStreamingUnitCountPerClusterList,
+			regionsSupportedInstanceType:           regionsSupportedInstanceType,
+			supportedKafkaInstanceTypesConfig:      &m.kafkaConfig.SupportedInstanceTypes.Configuration,
+			clusterService:                         m.clusterService,
+			dryRun:                                 !m.dataplaneClusterConfig.EnableDynamicScaleDownManagerScaleDownTrigger,
+			clusterID:                              clusterID,
+			indexesOfStreamingUnitForSameClusterID: existing.indexesOfStreamingUnitForSameClusterID,
+		}
+
+		glog.Infof("evaluating dynamic scale down for cluster '%s'", clusterID)
+		shouldScaleDown, err := dynamicScaleDownProcessor.ShouldScaleDown()
+		if err != nil {
+			errList.AddErrors(err)
+			continue
+		}
+		if shouldScaleDown {
+			glog.Infof("data plane scale down need detected for cluster '%s'", clusterID)
+			err := dynamicScaleDownProcessor.ScaleDown()
+			if err != nil {
+				errList.AddErrors(err)
+				continue
+			}
+		}
+	}
+
+	if errList.IsEmpty() {
+		return nil
+	}
+
+	return errList
+}
+
+func (m *DynamicScaleDownManager) logReconcileEventEnd() {
+	glog.Infoln("dynamic scale down reconcile event finished")
+}
```

Comment:
Why create a method that calls `Infoln` instead of using only `Infoln`?
I saw we did the same in `dynamic scale up mgr`: I believe I remember it was meant to be extended with some extra info, but that never happened.
My suggestion is to add these methods only when we have some logic for them.

Reply:
Thanks for the suggestion @ziccardi I've re-adjusted this part.

Reply:
Meaning to be extended was not the only reason. Aside from that, which might be useful:
1. Makes clear the intention of that logic. There are several places where creating a method that only includes one line is reasonable. See the ScaleUp manager for some cases where some conditions where added as 1 line methods to make intent more clear
1. Unconditionally log that line. By having removed that the behavior has been altered: Now when there's an error in the intermediate logic between where the defer originally was and the end of the method we won't perform logging. Whereas before it was unconditional. defer could be called with the glog infolfn call directly but again having it encapsulated as mentioned in the previous point seems reasonable to me.

Again I don't see any issue with having that even if there is no more than one line.

Reply:
> Now when there's an error in the intermediate logic between where the defer originally was and the end of the method we won't perform logging. 

I am trying to process this in context of https://github.com/machi1990/kas-fleet-manager/blob/9399c93d0c24692a61438c884ea8745af6446cee/internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go#L67-L75

Given that all the errors are accumulated and returned at the end, what errors do you think can cause the log from being processed?

Reply:
Thanks for confirming this. I'll change it there as well.